### ISSUE TITLE
Ability to list models on ollama.com

### ISSUE THREAD

**Issue by jkleckner** [ollama/ollama#286]
It would be nice to see what models are in the registry.  For example, `llama2:70b` is in there but not in the README.md

**Comment by technovangelist** [comment]
Yes, we are working on this. Soon there will be a site where you can see all this. We are still working on the right way to share this info.

**Comment by t3r-cni** [comment]
Is this https://ollama.ai/library? Issue closed?

**Comment by MassimilianoPasquini97** [comment]
@t3r-cni Would be nice to have a json version of library to fetch to, including model metadata such description, license and weight. Could be useful for third party developer and ollama cli with command like `ollama search` and `ollama show` for search and show detail of models.

**Comment by t3r-cni** [comment]
Thanks for the clarification. That would be nice.

**Comment by chrisweeksnz** [comment]
Here are a couple of quick bash/zsh functions that should do what you want, until a better system is in place (like a json endpoint for library):

```bash
function ollama_get_latest_model_tags(){
  # Find available models
    echo "Gathering available models..."
    OLLAMA_MODELS=$(\
      curl -s https://ollama.ai/library | \
      grep 'class="group"' | \
      sed -e 's|.*library\/||' \
          -e 's|".*||' \
          -e 's|\/.*||' | \
      sort \
      )

    # Finds tags of each model
    rm ${HOME}/.ollama_model_tag_library
    for MODEL in ${OLLAMA_MODELS}
    do 
      echo "Gathering available tags for ${MODEL}"
      curl -s https://ollama.ai/library/${MODEL}/tags | \
        grep "ollama pull" | \
        sed -e 's|<span .*Tag">||' \
            -e 's|<.*||' \
            -e 's|[ \t]*||' \
            -e 's|ollama pull ||' \
        >> ${HOME}/.ollama_model_tag_library
    done
}

function ollama_print_latest_model_tags(){
  cat ${HOME}/.ollama_model_tag_library
}

# Usage:
#
# Call ollama_get_latest_model_tags when you want to update the list of models and tags
ollama_get_latest_model_tags
#
# Call ollama_print_latest_model_tags to see a list of all models and tags. Use grep to find the model you desire.
ollama_print_latest_model_tags
#
# Please note that this will leave a single artifact on your Mac, a text file: ${HOME}/.ollama_model_tag_library
# You can delete this at any time, it will get recreated when/if you run ollama_get_latest_model_tags
```

This will be prone to breaking if the structure/content of the HTML changes substantially, but works at the moment... enjoy!

**Comment by JDRay42** [comment]
In the CLI, when someone enters 'ollama pull' without a model name, it should return a current list of available models.

**Comment by Sandros94** [comment]
> In the CLI, when someone enters 'ollama pull' without a model name, it should return a current list of available models.

I second this.
I'm building a small chat interface as a personal exercise and I wanted to simplify this process too

**Comment by actionproject-shree** [comment]
Taken from #7812 with a few modification, this can be achieved using a justfile
```
set positional-arguments := true

default:
  just --list

#list Ollama models
@models:
    curl -s https://ollama.com/library | grep -oP 'href="/library/\K[^"]+'

#list model tags
@tags model:
    curl -s https://ollama.com/library/$1/tags | grep -oP "/library/$1:\K.*(?=\")"

```

**Comment by actionproject-shree** [comment]
Taken from #7751  with a few modification, this can be achieved using a justfile
```
set positional-arguments := true

default:
  just --list

#list Ollama models
@models:
    curl -s https://ollama.com/library | grep -oP 'href="/library/\K[^"]+'

#list model tags
@tags model:
    curl -s https://ollama.com/library/$1/tags | grep -oP "/library/$1:\K.*(?=\")"

```
example output:
```
$ just tags llama3.3
latest
70b
70b-instruct-fp16
70b-instruct-q2_K
70b-instruct-q3_K_M
70b-instruct-q3_K_S
70b-instruct-q4_0
70b-instruct-q4_1
70b-instruct-q4_K_M
70b-instruct-q4_K_S
70b-instruct-q5_0
70b-instruct-q5_1
70b-instruct-q5_K_M
70b-instruct-q6_K
70b-instruct-q8_0

```
to setup just, which is similar to make but intended as a command runner, more on installation here: [just](https://github.com/casey/just)
just touch a **justfile** with contents above in the project root or home, the _just_ command looks for a justfile in the current shell working directory

if using docker, possible to pull _just_ in the ollama docker container with a _justfile_ already setup

**Comment by jmorganca** [comment]
This is now possible on [ollama.com](https://ollama.com)

**Comment by edwinjhlee** [comment]
> This is now possible on [ollama.com](https://ollama.com)

I have searched the repo and didn't find the clue.

Is there some related documentation published ? Or perhaps an curl uasge example ?

Thank you ~

**Comment by khimaros** [comment]
i don't think this issue has been addressed for CLI

**Comment by edwinjhlee** [comment]
> i don't think this issue has been addressed for CLI

I am only looking for a solution in API form.

https://github.com/ollama/ollama/issues/1070

**Comment by mtribiere** [comment]
I'm also looking for that API endpoint that would be really convenient.
Is there documentation about this somewhere?

**Comment by craigahobbs** [comment]
I, too, would like such a JSON API. It doesn't need to be live - it can be a static resource updated daily, making it inexpensive to host.

I've created such a resource for the [ollama-chat](https://github.com/craigahobbs/ollama-chat) model download page here:

[models.json](https://craigahobbs.github.io/ollama-chat/models/models.json)

The file format is available here:

[models.json File Format](https://craigahobbs.github.io/ollama-chat/api.html#var.vName='OllamaChatModels')

The `models.json` file is updated nightly from https://ollama.com/library. Ideally, the format would contain the quantization (Q), but that information isn't on the main page, and I wanted to keep my scraper simple.

The scraper approach works as a temporary measure, but it's fragile and not a long-term solution. An official Ollama API would be much better.

Is this what other people are thinking?

**Comment by Tickloop** [comment]
Is **ollama.com** open sourced? 

Its clear that web-scrapper based solutions can be built but a long term stable solution requires a JSON API. **ollama.com** seems to use Go+HTMX stack (best guess) which means a new endpoint will need to be created to support a JSON AtPI. 

Can the maintainers from **ollama.com** clarify if/how we can work on adding this functionality?
So either **ollama.com** will not support a JSON API for model registry or they do add an API upon which new cli commands may be built. 

I have a search client written in python: https://github.com/Tickloop/silsila/blob/main/src/search_ollama.py
that I can port as a cobra command.

If **ollama.com** starts supporting a JSON API we can refactor later. This has gone too long without a solution 😞

**Comment by i-am-mike-davis** [comment]
> Is **ollama.com** open sourced?
> 
> Its clear that web-scrapper based solutions can be built but a long term stable solution requires a JSON API. **ollama.com** seems to use Go+HTMX stack (best guess) which means a new endpoint will need to be created to support a JSON API.
> 
> Can the maintainers from **ollama.com** clarify if/how we can work on adding this functionality? So either **ollama.com** will not support a JSON API for model registry or they do add an API upon which new cli commands may be built.
> 
> I have a search client written in python: https://github.com/Tickloop/silsila/blob/main/src/search_ollama.py that I can port as a cobra command.
> 
> If **ollama.com** starts supporting a JSON API we can refactor later. This has gone too long without a solution 😞

I am not an **ollama.com** maintainer, but I made [ollama-admin-ui](https://github.com/i-am-mike-davis/ollama-admin-ui). 

While the project is a webapp that implements the aforementioned web-scraping solution, it comes preloaded with the ollama model/tag library catalog and utility classes which can be imported into other tools and the catalog dumped to JSON without running the web application or conducting a scraping operation. More details in the project README.

**Comment by anirbanbasu** [comment]
While intended for a different purpose, the project [ollama-downloader](https://github.com/anirbanbasu/ollama-downloader), which I maintain, now has a [`list-models`](https://github.com/anirbanbasu/ollama-downloader?tab=readme-ov-file#list-models) and [`list-tags`](https://github.com/anirbanbasu/ollama-downloader?tab=readme-ov-file#list-tags) capability. Both of these use web scraping.

**Comment by ashupednekar** [comment]
Hi, I understand it can be done by curling and grepping the API... but it would be useful to have an `ollama browse` command..

I'd be happy to take that up and raise a PR...  if it aligns with the goals

**Comment by ashupednekar** [comment]
is scraping the site really the only way? I get that it's SSR... aren't there any api's that can be used to list the models?

**Comment by Tickloop** [comment]
For new folk discovering this issue: 

PR Open for 5 months

https://github.com/ollama/ollama/pull/10046

No comment/feedback.

**Comment by Dutcho** [comment]
I expected the 'remote list' function. And now it's absent, I would very much appreciate if it gets added (incl. tags etc.).
Ideally, it would not be *only* CLI, but also Python SDK. But if the server exposes a JSON API, that's easily circumvented now and fixed later.

Re:
> No comment/feedback.

there's progress since that was posted.

Quite recently, @reid-endeavor approved with 'lgtm'. However, the branch has conflicts to be resolved before merging.
Perhaps @Tickloop you could find the time to look into those conflicts, so this ***can*** be merged, and we all can happily use it?

**Comment by Tickloop** [comment]
@Dutcho conflicts resolved. Good to merge. Still waiting on a review (been over a year)
